python/lsst/daf/butler/core/config.py

```diff
@@ -349,32 +406,102 @@ def names(self, topLevelOnly=False):
             if isinstance(d, collections.Sequence):
                 theseKeys = range(len(d))
             else:
                 theseKeys = d.keys()
             for key in theseKeys:
                 val = d[key]
-                levelKey = f"{base}.{key}" if base is not None else key
+                levelKey = base + (key,) if base is not None else (key,)
                 keys.append(levelKey)
                 if isinstance(val, (collections.Mapping, collections.Sequence)) and not isinstance(val, str):
-                    getKeys(val, keys, levelKey)
+                    getKeysAsTuples(val, keys, levelKey)
         keys = []
-        getKeys(self.data, keys, None)
+        getKeysAsTuples(self.data, keys, None)
         return keys
 
+    def names(self, topLevelOnly=False, delimiter=None):
+        """Get the delimited name of all the keys in the hierarchy.
+
+        The values returned from this method are guaranteed to be usable
+        to access items in the configuration object.
+
+        Parameters
+        ----------
+        topLevelOnly : `bool`, optional
+            If False, the default, a full hierarchy of names is returned.
+            If True, only the top level are returned.
+        delimiter : `str`, optional
+            Delimiter to use when forming the keys.  The delimiter must
+            not be present in any of the keys.  If `None` given the delimiter
+            will be automatically provided.
+
+        Returns
+        -------
+        names : `list`
+            List of all names present in the `Config`.
+
+        Notes
+        -----
+        This is different than the built-in method `dict.keys`, which will
+        return only the first level keys.
+
+        Raises
+        ------
+        ValueError
+            The externally specified delimiter is unsuitable because a key
+            clashes with the value.
+        """
+        if topLevelOnly:
+            return list(self.keys())
+
+        # Get all the tuples of hierarchical keys
+        nameTuples = self.nameTuples()
+
+        # Form big string for easy check of delimiter clash
+        combined = "".join("".join(str(s) for s in k) for k in nameTuples)
+
+        if delimiter is not None:
+            if delimiter in combined:
+                # For better error message we need to report the clashing key
+                def findMatch(names, searchStr):
+                    for n in names:
+                        for s in n:
+                            if searchStr in s:
+                                return s
+                    return None
+
+                raise ValueError(f"Specified delimiter, {delimiter!r} can not be used.  "
+                                 f"Clashes with key {findMatch(nameTuples, delimiter)!r}")
+        else:
+            # Start with something
+            delimiter = self._D
+            ntries = 0
+            while delimiter in combined:
+                log.debug(f"Trying delimiter '{delimiter}'")
```

Comment:
The message here isn't quite accurate; `delimiter` was already tried and found wanting.

Reply:
Good point. I had meant to remove that debug message anyhow.